yellow-paper/docs/contract-deployment/classes.md

```diff
@@ -16,114 +16,158 @@ Read the following discussions for additional context:
-| `artifact_hash` | `Field` | Hash of the entire contract artifact, including compiler information, proving backend, and all generated ACIR and Brillig. The specification of this hash is left to the compiler and not enforced by the protocol. |
-| `constructor_function` | [`PrivateFunction`](#private-function) | PublicFunction | Constructor for instances of this class. |
-| `private_functions` | `PrivateFunction[]` | List of private functions. |
-| `public_functions` | [`PublicFunction[]`](#public-function) | List of public functions. |
-| `unconstrained_functions` | [`UnconstrainedFunction[]`](#unconstrained-function) | List of unconstrained functions. |
-
-<!--
-TODO(palla): Do we need the artifact hash, if we're including the artifact hash of each individual function?
-A(mike): An artifact will include 'global' contract data which isn't contained within a function's artifact hash. E.g. state variable declarations; the number of functions in the contract; the compiler version used; etc.
-Q from mike: do we need the artifact hash in each individual function, if we have a 'global' artifact hash within the contract class struct? Does it make validation of code (before executing that code, for safety) faster? Hmm I suppose a per-function artifact hash is important, e.g. to link a function's combined acir & brillig bytecode to a vkhash (since the vkhash actually only represents the acir code).
-
-Q(mike): What should `constructor_function` be if there is no constructor? A bunch of `0` fields?
--->
-<!-- NOTE(palla): I'm deliberately omitting the portal bytecode hash here. -->
+| `artifact_hash` | `Field` | Hash of the contract artifact. The specification of this hash is not enforced by the protocol. Should include commitments to unconstrained code and compilation metadata. Intended to be used by clients to verify that an off-chain fetched artifact matches a registered class. |
+| `private_functions` | [`PrivateFunction[]`](#private-function) | List of individual private functions, constructors included. |
+| `packed_bytecode` | `Field[]` | [Packed bytecode representation](../public-vm/bytecode-validation-circuit.md#packed-bytecode-representation) of the AVM bytecode for all public functions in this contract. |
+
+Note that individual public functions are not first-class citizens in the protocol, so the contract entire bytecode is stored in the class, unlike private or unconstrained functions which are differentiated individual circuits recognized by the protocol.
+
+As for unconstrained functions, these are not used standalone within the protocol. They are either inlined within private functions, or called from a PXE as _getters_ for a contract. Calling from a private function to an unconstrained one in a different contract is forbidden, since the caller would have no guarantee of the code run by the callee. Considering this, unconstrained functions are not part of a contract class at the protocol level.
```

Comment:
> What do we do in our tests in the codebase at the moment, for a recipient to learn of a contract's unconstrained bytecode? Broadcast the unconstrained bytecode at deployment, or 'cheat'?

We cheat, of course. The deployment broadcasts only public function code.

> A suggestion for a useful 'rule of thumb' - but open to debate

I like it. This would mean allowing the registerer to broadcast a subset of unconstrained functions, but it makes sense. I wouldn't _enforce_ it in the registerer contract though, since it feels the valid selectors for `compute_note_hash_and_nullifier` may change quickly over time.

Reply:
Added a `broadcast_unconstrained_function` (note the singular), and its `private` counterpart for the sake of completeness, to the Registerer contract:

```
function broadcast_unconstrained_function(
  contract_class_id: Field,
  artifact_metadata: Field,
  private_functions_artifact_tree_root: Field,
  function_leaf_sibling_path: Field,
  function: { selector: Field, metadata: Field, bytecode: Field[] }[],
)
  emit_event ClassUnconstrainedFunctionBroadcasted(
    contract_class_id,
    artifact_metadata,
    unconstrained_functions_artifact_tree_root,
    function_leaf_sibling_path: Field,
    function,
  )
```